add cyphal (uavcan v1) support

PonomarevDA opened the pull request:
This PR adds a minimal implementation of Cyphal (uavcan v1.0) for ESC controlling.
It has the following required for any application functions:
- Node heartbeat (the node publishes uavcan.node.Heartbeat)
- Generic node information (the node response on uavcan.node.GetInfo.1.0 request)
- Register interface (response on uavcan.register.Access.1.0 and uavcan.register.List.1.0)

I also implemented templates for the following data types, but they do nothing yet. They might be completed if someone needs them or even removed for a while:
- uavcan.node.port.List (sends all supported port id)
- uavcan.node.ExecuteCommand (handles requests on restart or save parameters commands)

It also interacts with all necessary for ESC controlling data types, such as:
- reg.udral.service.actuator.common.sp.Vector4.0.1
- reg.udral.service.common.Readiness.0.1
- reg.udral.physics.electricity.PowerTs_0_1
- reg.udral.service.actuator.common.Feedback_0_1
- reg.udral.service.actuator.common.Status_0_1
- reg.udral.physics.dynamics.rotation.PlanarTs_0_1

All ESC related data types have non-fixed subject identifier. It is handled by the register interface that is based on AP_Param library. Since the implementation doesn't support automatically port id enumeration, you need to either manually or using a script from external server set their id (for ardupilot and for esc's).

Typical set of parameters and corresponded registers might be as on the picture below: 
![ardupilot_params](https://user-images.githubusercontent.com/36133264/160892348-bdcf6122-24b3-4ac9-b768-44970b7c4562.png)

I also forwarded esc telemetry to the corresponded MAVLink message:
![esc_telem](https://user-images.githubusercontent.com/36133264/160892583-821a31f1-431a-4174-bb10-59983c528031.png)

The implementation is based on `libcanard`, `o1heap`, `public_regulated_data_types` repositories which are added as submodules. It also uses `nunavut` for generating the code from DSDL. I tried to implement it in the similar way as AP_UAVCAN is implemented. You may use both versions of UAVCAN simultaneously, but on different CAN buses only.

According to the build summary comparison, on CUAV v5 the current branch used 14244 bytes of flash memory more than master branch. Although free flash on this branch on this hardware is 317088 bytes.

The firmware was tested on real copter based on cuav v5. I attached here a few flight examples [first](https://youtu.be/wBNh1XV1EMQ) and [second](https://youtu.be/FE_kxwhicWM) and corresponded [log files](https://drive.google.com/drive/folders/11hBPpcgdHG1a3gUMJ8z9DtMmeTAbuwBR?usp=sharing). I used Holybro kotleta20 ESCs based on sapog v3 firmware. Here should be noticed that I tested it before merging master branch into this branch. I may provide more tests in future. I also tried `Tools/scripts/build_all.sh` and it's ok.

I also want to notice that even if it was tested it is a still just a first version of implementation. It has a few assumptions, for example:
- it is expected to use it only for quadcopter since it uses exactly 4 ESCs,
- it works only on first CAN bus now.

I didn't test it on other hardware.

Any feedback is welcome.

PonomarevDA commented:
The problems with tests are related to proper installation of nunavut.
Typically, it might be installed as `python3 -m pip install pydsdl==1.13.0 nunavut`.
It also looks like `install-prereqs-ubuntu.sh` uses a system version of python, but nunavut requires at least 3.6 (3.7 is better). I added these packages to `PYTHON_PKGS` variable, but it doesn't work on workflow. I also don't see where this script is called from workflow.
So, I would be appreciated, if somebody help me how to solve this problem in more convenient way.

khancyr commented on .gitmodules:
```diff
@@ -34,3 +34,15 @@
+	url = https://github.com/UAVCAN/libcanard.git
+[submodule "modules/uavcan_v1/o1heap"]
+	path = modules/uavcan_v1/o1heap
+	url = https://github.com/pavel-kirienko/o1heap.git
```
why do we need this ? doesn't UAVCAN have this on its own repo ?

PonomarevDA replied:
No. Here we have the original repository of o1heap. There is nothing same inside uavcan organization. This repo link I got from the example from the libcanard readme.

khancyr commented on .gitmodules:
```diff
@@ -34,3 +34,15 @@
+	url = https://github.com/UAVCAN/public_regulated_data_types.git
+[submodule "modules/uavcan_v1/nunavut_c_headers"]
+	path = modules/uavcan_v1/nunavut_c_headers
+	url = https://github.com/PonomarevDA/nunavut_c_headers.git
```
we probably won't want to use a private repo for this.

PonomarevDA replied:
I could move this repo in any organization if it is acceptable. Although I've tried the approach that you suggested.

khancyr commented on Tools/environment_install/install-prereqs-arch.sh:
```diff
@@ -26,7 +26,7 @@ BASE_PKGS="base-devel ccache git gsfonts tk wget gcc"
 PX4_PKGS="lib32-glibc zip zlib ncurses"
 
-PYTHON_PKGS="future lxml pymavlink MAVProxy pexpect argparse matplotlib pyparsing geocoder pyserial empy"
+PYTHON_PKGS="future lxml pymavlink nunavut MAVProxy pexpect argparse matplotlib pyparsing geocoder pyserial empy"
```
can we add this as submodule and call it directly?

PonomarevDA replied:
fixed

khancyr commented on Tools/environment_install/install-prereqs-ubuntu.sh:
```diff
@@ -5,7 +5,6 @@ set -x
 
 if [ $EUID == 0 ]; then
     echo "Please do not run this script as root; don't sudo it!"
-    exit 1
```
why remove this ? that is wrong

PonomarevDA replied:
fixed

khancyr commented on Tools/environment_install/install-prereqs-ubuntu.sh:
```diff
@@ -267,6 +266,7 @@ if [ ${RELEASE_CODENAME} == 'xenial' ] || [ ${RELEASE_CODENAME} == 'disco' ] ||
 fi
 $PIP install --user -U $PYTHON_PKGS
 
+
```
oups, please remove

PonomarevDA replied:
fixed

khancyr commented on Tools/scripts/build_ci.sh:
```diff
@@ -246,19 +246,19 @@ for t in $CI_BUILD_TARGET; do
 
     if [ "$t" == "stm32f7" ]; then
         echo "Building mRoX21-777/"
-        $waf configure --Werror --board mRoX21-777
```
we need to keep --Werror on CI

PonomarevDA replied:
So, does it mean I should I try to come up with solution on libcanard side? I don't see what else can I fix on ardupilot side.

magicrub replied:
Why are there warnings?

PonomarevDA replied:
Sorry for the long answer. If disabling --Werror is unacceptable and we can't do something on the ardupilot side, so it might be better to move this discussion to the libcanard side. I've implemented a simple solution and opened a pull request: https://github.com/OpenCyphal/libcanard/pull/197.

PonomarevDA replied:
fixed on the libcanard side

khancyr commented on libraries/AP_UAVCAN_V1/AP_UAVCAN_V1.cpp:
```diff
@@ -0,0 +1,405 @@
+        publisher_manager.process_all();
+
+        static uint32_t next_pub_time_ts = 1000;
+        if (AP_HAL::millis() > next_pub_time_ts) {
```
why publish once a seconds ? just publish when you get a new message

PonomarevDA replied:
fixed

khancyr commented on libraries/AP_UAVCAN_V1/AP_UAVCAN_V1_esc.cpp:
```diff
@@ -0,0 +1,353 @@
+
+    WITH_SEMAPHORE(SRV_sem);
+
+    uint8_t src_idx = 0;
```
generally speaking, variable should be created on the scope of their usage.
here src_idx is only used on the for loop, so it should belong to the for loop only.

PonomarevDA replied:
fixed

khancyr reviewed (changes requested):
I add a few comments.
This will need some discussion on how to integrate this releatively to DroneCan

For the nunavut addition to install-prereqs-ubuntu.sh, that is normal that it doesn't work with CI as CI don't use install-prereqs-ubuntu.sh to not waste time on installation.
it would be best to call the dsls from a submodule as we are already doing with DroneCan.

That way we control the version of the transpiler we use.

PonomarevDA commented on libraries/AP_UAVCAN_V1/AP_UAVCAN_V1.cpp:
```diff
@@ -0,0 +1,405 @@
+void AP_UAVCAN_V1::loop(void)
+{
+    while (true) {
+        hal.scheduler->delay_microseconds(500); // it should be enough to handle max frame rate
```
I've tried to implement more similar to AP_UAVCAN way of blocking call  implementation.

Autonomost commented:
This will be great for my myxa esc's! I'll prioritize testing this on my vtol hardware asap. Awesome work!

PonomarevDA commented:
Since uavcan v1 and github organization with all corresponded modules was renamed into cyphal, I've changed the library name into AP_CYPHAL abd updated submodules.

PonomarevDA commented:
I've updated the library with the newest libcanard version and renamed somewhere uavcan to cyphal.

frontw commented:
Hi!
@Autonomost @khancyr Is there any news or forecasts on the testing of this PR?
Writing just to indicate the interest to have this feature in Ardupilot, as we are working on the design of a new state of the art (as far as we can) UAV. One of the key requirements is a modern unified on-board data bus with redundancy of sensors and actuators. 
Our customer, on the one hand, likes Ardupiliot and asks to base our solution on it, but, on the other hand, we should use the latest distributed on-board computing approaches, such as Cyphal. So the clarity on the possibility of this in Ardupilot, if this PR will be integrated, will determine the possibility of using it generally in our project.

mikkab commented:
Cool PR! just now I working for a new periphery using Cyphal protocol

PonomarevDA commented:
I've updated a bit the branch with changes that allows to easier add Cyphal sensors in future and fixed the conflicts.
I also made a test flight with the updates. The flight log is [here](https://drive.google.com/drive/u/1/folders/11hBPpcgdHG1a3gUMJ8z9DtMmeTAbuwBR). 

By the way, I'm currently working with [the branch](https://github.com/PonomarevDA/ardupilot/tree/pr-uavcan-v1-hitl) that has a few very first implementation of Cyphal sensors such as gps, magnetometer, baro and imu. Features of this PR + sensors from the mentioned branch are the minimal requirements for the full Cyphal HITL simulation. This tool might be useful for testing the firmware right now.
Here is [very first example](https://youtu.be/SDP_7tIml4A) of how the simulation process might look like: it has real motors and real ESCs, all internal board sensors are disabled and it uses fake cyphal sensors.
The documentation for the simulator is in process and I probably still need to make few minor fixes. I could write here when it will be ready for use and show more simulation examples.

Could you please advice what should I focus on next? Should I start with ardupilot wiki with description how to configure Cyphal. Or may be it's better to provide more flight/simulation tests?

wp0351034 commented:
您的邮件已经收到，尽快给您回复。北京堂仁翔科技有限公司联系人：王鹏  电话：135 2058 1985 ***@***.***网址：www.trxchn.comQQ:191896104地址：北京市海淀区花园北路14号环星大厦A座8层888室

SabadashAndrey commented:
Hi @tridge , please don't give us the cold shoulder.

Cyphal is a great library and has very flexible functionality. 

As Ardupilot users, we are focused on deep collaboration on motor controllers, servos to improve the power efficiency of our eVTOLs, it's important for us to have a deterministic system like Cyphal, on master branch.

Here good explanation why we need this:
https://forum.opencyphal.org/t/cyphal-vs-dronecan/1814

Please give due consideration to this pull request

frontw commented:
Great to see activity on this PR! 
Btw, at the moment we actually have Cyphal support on a basic set of our RaccoonLab onboard devices, which we hope to integrate with Ardupilot: 
- CAN-PWM converter;
- GNSS+Mag+Baro sensor;
- Airspeed sensor.
Links: [docs](https://raccoonlabdev.github.io/docs/), [store](https://raccoonlab.co/store).

tridge commented:
I'm really reluctant to take on Cyphal. The configuration is very silly, with a huge number of parameters, and even more as we add GPS, compass, baro, airspeed etc etc. If you get a parameter wrong then as I understand it you end up with network casts of structures, so a GPS packet gets interpreted as a compass etc
It will almost certainly break as we move the mainline ArduPilot DroneCAN code to libcanard (see #22563)

tridge commented:
we could do cyphal in lua, like we've done many other proprietary CAN protocols in lua, including DJI gimbals, EFI drivers etc

PonomarevDA commented:
Hi. Thanks for the feedback.
I will give an example of a cyphal network configuration based on this branch + gps, magnetometer, barometer and possibly other sensors. This should clear things up.

Meijin78 commented:
Cool PR! The new version fixes a number of architectural problems

PonomarevDA commented:
I've added port.List and *.type registers to my [branch with sensor implementations](https://github.com/PonomarevDA/ardupilot/tree/cyphal), and finished with the video example of Cyphal network configuration.

In [Cyphal network configuration for Ardupilot](https://youtu.be/xvcuNHGmXlg) video, I wanted to show that:
1. Port ID configuration is a fairly simple process. You don't have to interact directly with the autopilot parameters, so you don't have to worry about their amount.
2. Yukon allows you to prevent misconfiguration. By looking at the graph, it is easy to verify the correctness of the network. It should also be easy to implement automatic network correctness checking on the Yukon side. Just compare the .type registers of the nodes and warn if they are different.

Please enable the subtitles. The video is intended only for those who are already familiar with the basics of Cyphal. For those who are not, it may be necessary to create another video with more comments.

PonomarevDA commented:
For those who don't have Cyphal hardware, you can now try ArduPilot in Cyphal HITL mode.
I've implemented a simple JSON to Cyphal converter. Using the [ArduPilot Gazebo plugin](https://ardupilot.org/dev/docs/sitl-with-gazebo.html#sitl-with-gazebo) and an autopilot with firmware on the [cyphal-hitl branch](https://github.com/PonomarevDA/ardupilot/tree/cyphal-hitl) (it is based on this branch + gnss, mag, baro, imu, rangefinder and a few fixes) you can run a Cyphal HITL simulation.
I've made [a short video demonstration](https://youtu.be/-ij5IKP6kHg). Details on how to run it can be found in [the repository](https://github.com/RaccoonlabDev/libcanard_cyphal_application/tree/devel/examples/autopilot_hitl).
I hope this application will help with testing of this PR.

tridge commented:
@PonomarevDA and @pavel-kirienko I've opened a discussion on our forums to discuss the issues with cyphal:
https://discuss.ardupilot.org/t/discussion-of-cyphal-support/98016

tridge commented:
as discussed on this post:
https://discuss.ardupilot.org/t/discussion-of-cyphal-support
we won't be merging this, closing
